openmls/src/group/core_group/validation.rs

```diff
@@ -137,14 +137,31 @@ impl CoreGroup {
             if !signature_key_set.insert(signature_key) {
                 return Err(ProposalValidationError::DuplicateSignatureKeyAddProposal);
             }
-            let public_key = add_proposal
+
+            let proposal_init_key = add_proposal
                 .add_proposal()
                 .key_package()
                 .hpke_init_key()
                 .as_slice()
                 .to_vec();
+            let proposal_encryption_key = add_proposal
+                .add_proposal()
+                .key_package()
+                .leaf_node()
+                .encryption_key();
+
+            // ValSem113
+            if proposal_init_key == proposal_encryption_key.as_ref() {
+                return Err(ProposalValidationError::InitEncryptionKeyCollision);
+            }
+
             // ValSem102
-            if !public_key_set.insert(public_key) {
+            if !init_key_set.insert(proposal_init_key) {
+                return Err(ProposalValidationError::DuplicatePublicKeyAddProposal);
+            }
```

Comment:
I don't think this check is strictly necessary in that it is not mandated by the spec. If I understand Section 10.1 correctly, it is technically okay for two KeyPackages to have the same init key. No need to remove it with this PR, though I guess.